core/src/main/java/io/grpc/Metadata.java

```diff
@@ -134,15 +135,20 @@ public Integer parseAsciiString(String serialized) {
   private final Map<String, List<MetadataEntry>> store =
       new LinkedHashMap<String, List<MetadataEntry>>();
 
+  /** The number of headers stored by this metadata.  */
+  private int storeCount;
```

Comment:
Interestingly, this is incorrect, because the getAll() method can allow the consumer to call remove, allowing this to get out of sync.

Not sure if we want to allow that, or add in code to fix things up after iteration.  Either way, the code without this change is wrong, because if someone does call remove on that iterator, it could leave our code with a key with an empty list in `store`.